File: app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionPresenter.java
package org.dhis2.usescases.programStageSelection;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import org.dhis2.data.schedulers.SchedulerProvider;
import org.dhis2.utils.Result;
import org.dhis2.utils.RulesUtilsProvider;
import org.hisp.dhis.android.core.program.ProgramStage;
import org.hisp.dhis.rules.models.RuleEffect;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

import static androidx.annotation.VisibleForTesting.PRIVATE;

/**
 * QUADRAM. Created by ppajuelo on 31/10/2017.
 */

public class ProgramStageSelectionPresenter implements ProgramStageSelectionContract.Presenter {

    private final RulesUtilsProvider ruleUtils;
    private final SchedulerProvider schedulerProvider;
    private ProgramStageSelectionContract.View view;
    public CompositeDisposable compositeDisposable;
    private ProgramStageSelectionRepository programStageSelectionRepository;

    public ProgramStageSelectionPresenter(ProgramStageSelectionContract.View view, ProgramStageSelectionRepository programStageSelectionRepository, RulesUtilsProvider ruleUtils, SchedulerProvider schedulerProvider) {
        this.view = view;
        this.programStageSelectionRepository = programStageSelectionRepository;
        this.ruleUtils = ruleUtils;
        this.schedulerProvider = schedulerProvider;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onBackClick() {
        if (view != null)
            view.back();
    }

    @Override
    public void getProgramStages(String programId, @NonNull String uid) {

        Flowable<List<ProgramStage>> stagesFlowable = programStageSelectionRepository.enrollmentProgramStages(programId, uid);
        Flowable<Result<RuleEffect>> ruleEffectFlowable = programStageSelectionRepository.calculate();

        // Combining results of two repositories into a single stream.
        Flowable<List<ProgramStage>> stageModelsFlowable = Flowable.zip(
                stagesFlowable.subscribeOn(schedulerProvider.io()),
                ruleEffectFlowable.subscribeOn(schedulerProvider.io()),
                this::applyEffects);

        compositeDisposable.add(stageModelsFlowable
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        view::setData,
                        Timber::e));
    }

    @VisibleForTesting()
    public List<ProgramStage> applyEffects(List<ProgramStage> stageModels, Result<RuleEffect> calcResult) {
        if (calcResult.error() != null) {
            Timber.e(calcResult.error());
            return stageModels;
        }

        Map<String, ProgramStage> stageView = toMap(stageModels);

        ruleUtils.applyRuleEffects(stageView, calcResult);

        return new ArrayList<>(stageView.values());
    }

    @NonNull
    private static Map<String, ProgramStage> toMap(@NonNull List<ProgramStage> stageViews) {
        Map<String, ProgramStage> map = new LinkedHashMap<>();
        for (ProgramStage stageModel : stageViews) {
            map.put(stageModel.uid(), stageModel);
        }
        return map;
    }

    @Override
    public void onDettach() {
        compositeDisposable.clear();
    }

    @Override
    public void displayMessage(String message) {
        view.displayMessage(message);
    }

    @Override
    public void onProgramStageClick(ProgramStage programStage) {
        view.setResult(programStage.uid(), programStage.repeatable(), programStage.periodType());
    }

    @Override
    public int getStandardInterval(String programStageUid) {
        return programStageSelectionRepository.getStage(programStageUid).standardInterval() != null ?
                programStageSelectionRepository.getStage(programStageUid).standardInterval() :
                0;
    }
}

File: app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionPresenter.kt
package org.dhis2.usescases.programStageSelection

import androidx.annotation.VisibleForTesting
import io.reactivex.Flowable
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.dhis2.commons.resources.D2ErrorUtils
import org.dhis2.commons.resources.MetadataIconProvider
import org.dhis2.commons.schedulers.SchedulerProvider
import org.dhis2.commons.viewmodel.DispatcherProvider
import org.dhis2.form.data.RulesUtilsProvider
import org.dhis2.form.model.EventMode
import org.dhis2.tracker.events.CreateEventUseCase
import org.hisp.dhis.android.core.program.ProgramStage
import org.hisp.dhis.mobile.ui.designsystem.theme.SurfaceColor
import org.hisp.dhis.rules.models.RuleEffect
import timber.log.Timber

class ProgramStageSelectionPresenter(
    private val view: ProgramStageSelectionView,
    private val programStageSelectionRepository: ProgramStageSelectionRepository,
    private val ruleUtils: RulesUtilsProvider,
    private val metadataIconProvider: MetadataIconProvider,
    private val schedulerProvider: SchedulerProvider,
    private val dispatcher: DispatcherProvider,
    private val createEventUseCase: CreateEventUseCase,
    private val d2ErrorUtils: D2ErrorUtils,
) {
    var compositeDisposable: CompositeDisposable = CompositeDisposable()

    fun onBackClick() {
        view.back()
    }

    fun programStages() {
        val stagesFlowable = programStageSelectionRepository.enrollmentProgramStages()
        val ruleEffectFlowable = programStageSelectionRepository.calculate()

        // Combining results of two repositories into a single stream.
        val stageModelsFlowable =
            Flowable.zip(
                stagesFlowable.subscribeOn(schedulerProvider.io()),
                ruleEffectFlowable.subscribeOn(schedulerProvider.io()),
            ) { stageModels, calcResult ->
                applyEffects(
                    stageModels,
                    calcResult,
                )
            }
        compositeDisposable.add(
            stageModelsFlowable
                .map { programStages ->
                    programStages.map { programStage ->
                        ProgramStageData(
                            programStage,
                            metadataIconProvider(programStage.style(), SurfaceColor.Primary),
                        )
                    }
                }.subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::handleProgramStages) { t: Throwable? ->
                    Timber.e(t)
                },
        )
    }

    private fun handleProgramStages(programStages: List<ProgramStageData>) {
        when (programStages.size) {
            1 ->
                view.setResult(
                    programStageUid = programStages.first().programStage.uid(),
                    repeatable = programStages.first().programStage.repeatable() == true,
                    periodType = programStages.first().programStage.periodType(),
                )

            else -> view.setData(programStages)
        }
    }

    @VisibleForTesting
    fun applyEffects(
        stageModels: List<ProgramStage>,
        calcResult: Result<List<RuleEffect>>,
    ): List<ProgramStage> {
        if (calcResult.isFailure) {
            Timber.e(calcResult.exceptionOrNull())
            return stageModels
        }
        val stageView = stageModels.associateBy { it.uid() }.toMutableMap()
        ruleUtils.applyRuleEffects(
            stageView,
            Result.success(calcResult.getOrDefault(emptyList())),
        )
        return ArrayList(stageView.values)
    }

    fun onDettach() {
        compositeDisposable.clear()
    }

    fun displayMessage(message: String?) {
        view.displayMessage(message)
    }

    fun onProgramStageClick(programStage: ProgramStage) {
        if (programStage.access().data().write()) {
            view.setResult(
                programStage.uid(),
                programStage.repeatable() == true,
                programStage.periodType(),
            )
        } else {
            view.displayMessage(null)
        }
    }

    fun getStandardInterval(programStageUid: String): Int =
        programStageSelectionRepository.getStage(programStageUid)?.standardInterval() ?: 0

    fun onOrgUnitForNewEventSelected(
        programStageUid: String,
        programUid: String,
        orgUnitUid: String,
        enrollmentUid: String?,
    ) {
        CoroutineScope(dispatcher.io()).launch {
            createEventUseCase(
                programUid = programUid,
                orgUnitUid = orgUnitUid,
                programStageUid = programStageUid,
                enrollmentUid = enrollmentUid,
            ).fold(
                onSuccess = { eventUid ->
                    view.goToEventDetails(
                        eventUid = eventUid,
                        eventMode = EventMode.NEW,
                        programUid = programUid,
                    )
                },
                onFailure = {
                    view.displayMessage(d2ErrorUtils.getErrorMessage(it))
                },
            )
        }
    }
}
